controller/src/main/java/io/pravega/controller/server/SegmentHelper.java

```diff
@@ -267,42 +269,28 @@ private String getTransactionName(String scope, String stream, long segmentId, U
+                                                          final long sourceSegmentId,
+                                                          final List<UUID> txId,
+                                                          final String delegationToken,
+                                                          final long clientRequestId) {
         Preconditions.checkArgument(getSegmentNumber(targetSegmentId) == getSegmentNumber(sourceSegmentId));
         final Controller.NodeUri uri = getSegmentUri(scope, stream, sourceSegmentId);
         final String qualifiedNameTarget = getQualifiedStreamSegmentName(scope, stream, targetSegmentId);
-        final String transactionName = getTransactionName(scope, stream, sourceSegmentId, txId);
-        final WireCommandType type = WireCommandType.MERGE_SEGMENTS;
+        final List<String> transactionNames = txId.stream().map(x -> getTransactionName(scope, stream, sourceSegmentId, x)).collect(Collectors.toList());
+        final WireCommandType type = WireCommandType.MERGE_SEGMENTS_BATCH;
 
         RawClient connection = new RawClient(ModelHelper.encode(uri), connectionPool);
         final long requestId = connection.getFlow().asLong();
         
-        WireCommands.MergeSegments request = new WireCommands.MergeSegments(requestId,
-                qualifiedNameTarget, transactionName, delegationToken);
+        WireCommands.MergeSegmentsBatch request = new WireCommands.MergeSegmentsBatch(requestId, qualifiedNameTarget, transactionNames, delegationToken);
 
         return sendRequest(connection, clientRequestId, request)
                 .thenCompose(r -> {
-                    handleReply(clientRequestId, r, connection, transactionName, WireCommands.MergeSegments.class, type);
-                    if (r instanceof WireCommands.NoSuchSegment) {
-                        WireCommands.NoSuchSegment reply = (WireCommands.NoSuchSegment) r;
-                        if (reply.getSegment().equals(transactionName)) {
-                            // idempotent case when segment is already merged, we get the write offset for the parent segment.
-                            return getSegmentInfo(scope, stream, targetSegmentId, delegationToken, clientRequestId)
-                                    .thenApply(WireCommands.StreamSegmentInfo::getWriteOffset);
-                        } else {
-                            log.warn(clientRequestId, "Commit Transaction: Source segment {} not found.",
-                                    reply.getSegment());
-                            return CompletableFuture.completedFuture(-1L);
-                        }
-                    } else {
-                        return CompletableFuture.completedFuture(((WireCommands.SegmentsMerged) r).getNewTargetWriteOffset());
-                    }
+                    handleReply(clientRequestId, r, connection, qualifiedNameTarget, WireCommands.MergeSegmentsBatch.class, type);
```

Comment:
this method will throw a `WireCommandFailedException(type, WireCommandFailedException.Reason.SegmentDoesNotExist)`  in case the SSS responds with a `NoSuchSegment` . Could you elaborate where this exception is being handled?

Reply:
In the withRetries() of StreamMetadataTasks.mergeTxnSegments()